### ISSUE TITLE
mouse scroll wheel events not passed to tmux

### ISSUE THREAD

**Issue by cjw296** [jumptrading/FluentTerminal#196]
If I use the chrome ssh app to ssh to my linux desktop, and then attach to a tmux session, I can use the mouse scroll wheel to scroll back in the tmux buffer history.

With Fluent, the scroll events don't appear to be passed to tmux.

(I think tmux called this stuff "mouse mode", if that's helpful)

**Comment by mjs** [comment]
Agreed that it would be interesitng to see if this can be made to work. Console mode Vim can also support mouse events (enabled with `set mouse=a` - https://vim.fandom.com/wiki/Using_the_mouse_for_Vim_in_an_xterm).

**Comment by cjw296** [comment]
To be clear: this is quite a blocker for me, I have to fall back to the Chrome SSH app if I want to use my tmux sessions.

**Comment by mjs** [comment]
@cjw296 until this is sorted out, you know about tmux's keyboard based scrolling features right?

https://superuser.com/questions/209437/how-do-i-scroll-in-tmux

**Comment by cjw296** [comment]
I've tried and failed to use keyboard-based scrolling a few times, my brain just doesn't work with keyboard-based scrolling, sadly :'(

**Comment by mjs** [comment]
> I've tried and failed to use keyboard-based scrolling a few times, my brain just doesn't work with keyboard-based scrolling, sadly :'(

Fair enough :)

**Comment by MaxRis** [comment]
This issue happens also with Hyper.

**Comment by cjw296** [comment]
Could we remove the "low priority" on this? It effectively means I can't use tmux with Fluent and have to go find another ssh client or stop using tmux.

**Comment by mjs** [comment]
> Could we remove the "low priority" on this? It effectively means I can't use tmux with Fluent and have to go find another ssh client or stop using tmux.

Done. It was marked that way to indicate that the stability issues at the time were more pressing.

**Comment by MaxRis** [comment]
That seems to be an issue with XTerm.JS or also combined with UWP WebView.
tmux in mouse mode with XTerm.JS works good on Ubuntu, but Hyper and VS Code on Windows have the same issues as Fluent.

**Comment by MaxRis** [comment]
ConPTY doesn't support mouse events yet. Support is in MSFT Console team's plans, but active work will not be started most of all till the end of this year https://github.com/microsoft/terminal/issues/376#issuecomment-535498018
Based on above referenced comment it should be possible to contribute into ConPTY project.

**Comment by MaxRis** [comment]
The initial core reason of the issue is not triggered Xterm.JS's `onData` method. It's triggered successfully on Linux (checked with Hyper terminal) in alt screen mode and sends something like `ESC64;70;12M` or `ESC65;70;12M` for mouse wheel scroll events.
On Windows Xterm.js sends nothing to WinPTY or ConPTY terminal sessions on mouse events.

**Comment by MaxRis** [comment]
It's turned out that XTerm.JS receives wheel scroll events well, but it doesn't send `onData` events, because it's expected to send them only under ALT screen buffer, but both ConPTY and WinPTY for some reasons doesn't send `[?1049h` CSI (request to save cursor state and switch to alternative screen mode) on time when `tmux` is launched.

Based on above the issue on Windows seems is much more wider: XTerm.JS doesn't know when terminal session should switch to ALT buffer and back.

For ConPTY it's known issue https://github.com/microsoft/terminal/issues/381
Trying to figure out what happens with WinPTY.

**Comment by MaxRis** [comment]
XTerm.JS doesn't send `onData` mouse wheel event because ALT buffer https://github.com/xtermjs/xterm.js/blob/ea922fe54853db34bcd3dda4f3da110f25905b46/src/Terminal.ts#L900 (which is without scrollback history) was not activated during `tmux` launch https://github.com/xtermjs/xterm.js/blob/ea922fe54853db34bcd3dda4f3da110f25905b46/src/InputHandler.ts#L1346 (`[?1049h` CSI was not received by XTerm.JS)

**Comment by daviesalex** [comment]
FWIW, while this wont be what Chris wants to hear, I have a hard time thinking that working on this makes much sense if Microsoft are working on it for their terminal, and have stated that fixing it inside ConPTY is "high on their backlog". If it can be made to work easily for WinPTY that would be great, but the end game is probably that Microsoft fix this in ConPTY and the other bugs that folk hit (e.g. with mosh) and everybody can use that.

**Comment by MaxRis** [comment]
@daviesalex right, I'm, definitely, going to focus on WinPTY and XTerm.JS for now.
MS guys declared that real work on mouse mode and related things in ConPTY will start only next year.

Need to mention that there are huge amount of complains that mouse mode usually doesn't work on Windows for most of available terminals.

Can confirm that it works for WSL terminal (but not from Fluent).

**Comment by MaxRis** [comment]
It's turned out that WinPTY doesn't support mouse mode by default.
There is an attempt to have mouse mode supported with calling of special WinPTY API ( It's not part of Winpty.NET NuGet package interface ) https://github.com/rprichard/winpty/blob/850661d02bd8b9e5b7c802da51cf47d1140f49dd/src/include/winpty.h#L102
Enabling of mouse mode turns on sending of expected SGR mouse wheel events, but propagation fails somewhere later (possibly again inside WinPTY) and tmux doesn't respond correctly.

**Comment by MaxRis** [comment]
Turned out that OpenSSH (ssh.exe) source code also should be adjusted to support converting of mouse input console events into escape sequences to be sent to remote host (currently, mouse events are just ignored by OpenSSH).

**Comment by MaxRis** [comment]
Possibility to have `automatic` support of mouse mode switching (CLI app sends escape sequences to turn on/off mouse mode) is blocked by high level issue related to Windows Console limitations https://github.com/jumptrading/FluentTerminal/issues/226

We still can implement mouse mode support turned on/off by user manually with keyboard shortcut, for example. Since System's OpenSSH is used, we should be able more easily extend `mosh` build with mouse events support. (both OpenSSH and mosh should start to support handling of mouse input event from Windows Console)
